native/src/binaryRoutePlanner.cpp

```diff
@@ -676,90 +670,121 @@ bool proccessRestrictions(RoutingContext* ctx, SHARED_PTR<RouteSegment>& segment
 	SHARED_PTR<RouteSegment> parent = getParentDiffId(segment);
 
 	if (!reverseWay && road->restrictions.size() == 0 &&
-		(parent.get() == NULL || parent->road->restrictions.size() == 0)) {
+		(parent.get() == NULL || parent->road == nullptr || parent->road->restrictions.size() == 0)) {
 		return false;
 	}
 	clearSegments(ctx->segmentsToVisitPrescripted);
 	clearSegments(ctx->segmentsToVisitNotForbidden);
 	processRestriction(ctx, inputNext, reverseWay, 0, road);
-	if (parent.get() != NULL) {
+	if (parent.get() != NULL && parent->road != nullptr) {
 		processRestriction(ctx, inputNext, reverseWay, segment->road->id, parent->road);
 	}
 	return true;
 }
 
 SHARED_PTR<RouteSegment> processIntersections(RoutingContext* ctx, SEGMENTS_QUEUE& graphSegments,
-											  VISITED_MAP& visitedSegments, double distFromStart,
-											  SHARED_PTR<RouteSegment>& segment, int segmentPoint,
-											  SHARED_PTR<RouteSegment>& inputNext, bool reverseWaySearch,
-											  bool doNotAddIntersections, bool* processFurther) {
-	bool thereAreRestrictions;
-	SHARED_PTR<RouteSegment> itself;
-	vector<SHARED_PTR<RouteSegment> >::iterator nextIterator;
-	if (inputNext.get() != NULL && inputNext->getRoad()->getId() == segment->getRoad()->getId() &&
-		inputNext->next.get() == NULL) {
-		thereAreRestrictions = false;
-	} else {
-		thereAreRestrictions = proccessRestrictions(ctx, segment, inputNext, reverseWaySearch);
-		if (thereAreRestrictions) {
-			nextIterator = ctx->segmentsToVisitPrescripted.begin();
-			if (TRACE_ROUTING) {
-				OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Debug, "  >> There are restrictions");
+											  VISITED_MAP& visitedSegments, SHARED_PTR<RouteSegment>& currentSegment,
+											  bool reverseWaySearch, bool doNotAddIntersections) {
+	SHARED_PTR<RouteSegment> nextCurrentSegment = nullptr;
+	int targetEndX = reverseWaySearch ? ctx->startX : ctx->targetX;
+	int targetEndY = reverseWaySearch ? ctx->startY : ctx->targetY;
+	const int x = currentSegment->getRoad()->pointsX[currentSegment->getSegmentEnd()];
+	const int y = currentSegment->getRoad()->pointsY[currentSegment->getSegmentEnd()];
+	float distanceToEnd = h(ctx, x, y, targetEndX, targetEndY);
+	// reassign @distanceToEnd to make it correct for visited segment
+	currentSegment->distanceToEnd = distanceToEnd;
+
+	SHARED_PTR<RouteSegment> connectedNextSegment = ctx->loadRouteSegment(x, y, reverseWaySearch);
+	SHARED_PTR<RouteSegment> roadIter = connectedNextSegment;
+	bool directionAllowed = true;
+	bool singleRoad = true;
+	while (roadIter != nullptr) {
+		if (currentSegment->getSegmentEnd() == roadIter->getSegmentStart() &&
+			roadIter->road->getId() == currentSegment->getRoad()->getId()) {
```

Comment:
Need same style here, please use getRoad()

Reply:
Better add check roadIter->getRoad() != nullptr
I am not sure that getRoad() can not be nullptr

Reply:
Only start and end segments can have RouteSegment with road == nullptr